File: scripts/nearest.R

library(GenomicRanges)
library(data.table)

fc = snakemake@input[["chip"]]
fb = snakemake@input[["background"]]


print("Reading data table")
cmd = paste0("zcat ", fc, " | cut -f 1-3,6")
cmd_bg= paste0("zcat ", fb, " | cut -f 1-3,6")
print(cmd)
chip_df = fread(cmd, header=FALSE, col.names=c("Chromosome", "Start", "End", "Strand"))
input_df = fread(cmd_bg, header=FALSE, col.names=c("Chromosome", "Start", "End", "Strand"))

print("creating granges")
chip = GRanges(seqnames = chip_df$Chromosome, ranges = IRanges(start = chip_df$Start, end = chip_df$End), strand=chip_df$Strand)
input = GRanges(seqnames = input_df$Chromosome, ranges = IRanges(start = input_df$Start, end = input_df$End), strand=input_df$Strand)

print("finding nearest")

start.time <- Sys.time()
result = nearest(chip, input, select="arbitrary")
result = chip[result]
end.time <- Sys.time()

time.taken <- end.time - start.time


time.taken <- as.numeric(time.taken, units="secs")

print(result)

write(time.taken, file=snakemake@output[["time"]])

write.table(result, file=snakemake@output[["result"]], sep=" ", row.names=FALSE, quote=FALSE)

File: scripts/nearest.py
import pandas as pd

import numpy as np
# import ray
# ray.init(num_cpus=2)

from pyranges import PyRanges

from time import time
import datetime

chip_f = snakemake.input.chip
background_f = snakemake.input.background

nrows = None

chip = pd.read_table(chip_f, sep="\t", usecols=[0, 1, 2, 5], header=None, nrows=nrows,
                     names="Chromosome Start End Strand".split(),
        dtype= {"Start": np.int32, "End": np.int32, "Chromosome": "category", "Strand": "category"})

start_init = time()
cgr = PyRanges(chip, copy_df=False)
end_init = time()

print(end_init - start_init)

background = pd.read_table(background_f, sep="\t", usecols=[0, 1, 2, 5], nrows=nrows,
                           header=None, names="Chromosome Start End Strand".split(),
        dtype= {"Start": np.int32, "End": np.int32, "Chromosome": "category", "Strand": "category"})

start_init = time()
bgr = PyRanges(background, copy_df=False)
end_init = time()

print(end_init - start_init)
start = time()
result = cgr.nearest(bgr, strandedness="same")

end = time()

print(result)

total = end - start

total_dt = datetime.datetime.fromtimestamp(total)

minutes_seconds = total_dt.strftime('%-M.%-S.%f')

open(snakemake.output[0], "w+").write(minutes_seconds)

# result.df.to_csv(snakemake.output.result, sep=" ", index=False)
